pkg/reporting/report.go

```diff
@@ -0,0 +1,251 @@
+	K8sVersion                string `json:"k8s_version"`
+	K8sDistribution           string `json:"k8s_distribution,omitempty"`
+	KotsVersion               string `json:"kots_version"`
+	KotsInstallID             string `json:"kots_install_id,omitempty"`
+	KurlInstallID             string `json:"kurl_install_id,omitempty"`
+	IsGitOpsEnabled           bool   `json:"is_gitops_enabled"`
+	GitOpsProvider            string `json:"gitops_provider"`
+	DownstreamChannelID       string `json:"downstream_channel_id,omitempty"`
+	DownstreamChannelSequence uint64 `json:"downstream_channel_sequence,omitempty"`
+	DownstreamChannelName     string `json:"downstream_channel_name,omitempty"`
+	DownstreamSequence        *int64 `json:"downstream_sequence,omitempty"`
+	DownstreamSource          string `json:"downstream_source,omitempty"`
+	InstallStatus             string `json:"install_status,omitempty"`
+	PreflightState            string `json:"preflight_state,omitempty"`
+	SkipPreflights            bool   `json:"skip_preflights"`
+	ReplHelmInstalls          int    `json:"repl_helm_installs"`
+	NativeHelmInstalls        int    `json:"native_helm_installs"`
+}
+
+func (i *InstanceReportEvent) GetReportType() string {
+	return "instance"
+}
+
+func (i *InstanceReportEvent) GetReportSecretName(appSlug string) string {
+	return fmt.Sprintf(ReportSecretNameFormat, fmt.Sprintf("%s-%s", appSlug, i.GetReportType()))
+}
+
+func (i *InstanceReportEvent) GetReportSecretKey() string {
+	return ReportSecretKey
+}
+
+func (i *InstanceReportEvent) GetReportEventLimit() int {
+	return ReportEventLimit
+}
+
+func (i *InstanceReportEvent) GetReportMtx() *sync.Mutex {
+	return &instanceReportMtx
+}
+
+type PreflightReport struct {
+	Events []PreflightReportEvent `json:"events"`
+}
+
+type PreflightReportEvent struct {
+	ReportedAt      int64  `json:"reported_at"`
+	LicenseID       string `json:"license_id"`
+	InstanceID      string `json:"instance_id"`
+	ClusterID       string `json:"cluster_id"`
+	Sequence        int64  `json:"sequence"`
+	SkipPreflights  bool   `json:"skip_preflights"`
+	InstallStatus   string `json:"install_status"`
+	IsCLI           bool   `json:"is_cli"`
+	PreflightStatus string `json:"preflight_status"`
+	AppStatus       string `json:"app_status"`
+	KotsVersion     string `json:"kots_version"`
+}
+
+func (p *PreflightReportEvent) GetReportType() string {
+	return "preflight"
+}
+
+func (p *PreflightReportEvent) GetReportSecretName(appSlug string) string {
+	return fmt.Sprintf(ReportSecretNameFormat, fmt.Sprintf("%s-%s", appSlug, p.GetReportType()))
+}
+
+func (p *PreflightReportEvent) GetReportSecretKey() string {
+	return ReportSecretKey
+}
+
+func (p *PreflightReportEvent) GetReportEventLimit() int {
+	return ReportEventLimit
+}
+
+func (p *PreflightReportEvent) GetReportMtx() *sync.Mutex {
+	return &preflightReportMtx
+}
+
+func CreateReportEvent(clientset kubernetes.Interface, namespace string, appSlug string, event ReportEvent) error {
```

Comment:
It was a bit confusing to me at first glance why the `ReportEvent` is the interface rather than the `Report`. What do you think of making `Report` be the interface, renaming this function to `UpsertReport`, and the caller can decide how many events to send (in our case, it'll be a single event for now)?